apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Logging/GoogleLogger.cs

```diff
@@ -194,7 +194,7 @@ public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Except
                 Severity = logLevel.ToLogSeverity(),
                 Timestamp = Timestamp.FromDateTime(_clock.GetCurrentDateTimeUtc()),
                 JsonPayload = jsonStruct,
-                Labels = { labels },
+                Labels = { labels.Where(kvp => !string.IsNullOrEmpty(kvp.Value)).ToDictionary(kvp => kvp.Key, kvp => kvp.Value) },
```

Comment:
I don't like silently removing labels that client code have set, even if their value is `null` and sending the label will make the request fail. I rather the request fail and have the client realize that they cannot send a `null` value and fix their code.
With the changes on the providers below, we make sure that none of our providers are sending labels with null values. Any other labels client code sets, either through options or custom providers, they should take care of not sending null values.

Reply:
I agree. However, the point of this PR was to make `GoogleLogger.Log()` never throw as that leads to all sorts of unexpected behavior. Also throwing from loggers is discouraged, see https://github.com/aspnet/AspNetCore/issues/3460#issuecomment-416447852 for reference. I once had an error that the username was not populated correctly so the Google logger failed, which caused my app to completely crash. Took me a while to figure out the root cause. That issue was the reason for me to create this PR.

So I'm not sure what the right approach is here. I agree that failing silently is bad and can be annoying, but the consequences of throwing exceptions might be worse. What do you think about this?

Reply:
I see this PR as more of, "let's make sure we don't introduce any Exception causing issues" than let's make sure no Exception is thrown ever, there are probably several other ways in which `GoogleLogger.Log()` can throw that have nothing to do with labels. But I do agree with "loggers should not throw" and the fact that when they do it's hard to diagnose. So, what do you think if we do the following:

- Don't remove the null labels.
- Let's wrap the whole of `GoogleLogger.Log()` in a try-catch and honour `_loggerOptions.RetryOptions.ExceptionHandling`, if it's set to `Ignore`, we swallow the Exception else we rethrow? Note that we don't have to honour the other Retry settings because for transient server faults `_consumer.Receive` will already do that and anyway, whatever Exceptions are thrown in `GoogleLogger.Log()` are not really transient so it makes no sense to retry. This approach will yield a different result from yours in that when there are null labels nothing will be logged, but I'm fine with that, I think users will find that more suspicious and can knowingly turn on the propagation of Exceptions to check what's happening.
- Document `ILogEntryLabelProvider` explicitly saying that null keys or values are not allowed.
- Document LoggerOptions.Labels explicitly saying that null keys or values are not allowed. Actually we can probably even check that there are no null labels and throw here since this Exception will be trown at app Startup and not when attempting to log.

What do you think?

Reply:
That sounds good. I will look at implementing this approach.